### DISCUSSION TITLE
Make principle:keep source code readily comprehensible

### DISCUSSION THREAD

**Discussion by athinboy** [dotnet/csharplang#4522] [General]
First,apologize for my english.
Last several years,  csharplang had added more and more feature . But as a open source program language, source code should be  readily comprehensible ,or it will hamper other people  effort to contribute to open source project .

For example,when I study code of dotnet/diagnostics,It took me lots of time to understand CollectCommandHandler.CollectCommand method that why the first arguments is a ICommandHandler  until I find  CommandExtensions.Add method.
From my personnal oponion, i do not agree with this feature . It make source code too complex to understand . 
```
        public static Command CollectCommand() =>
            new Command(
                name: "collect",
                description: "Collects a diagnostic trace from a currently running process or launch a child process and trace it. Append -- to the collect command to instruct the tool to run a command and trace it immediately. When tracing a child process, the exit code of dotnet-trace shall be that of the traced process unless the trace process encounters an error.") 
            {
                // Handler
                HandlerDescriptor.FromDelegate((CollectDelegate)Collect).GetCommandHandler(),
                // Options
                CommonOptions.ProcessIdOption(),
                CircularBufferOption(),
                OutputPathOption(),
                ProvidersOption(),
                ProfileOption(),
                CommonOptions.FormatOption(),
                DurationOption(),
                CLREventsOption(),
                CLREventLevelOption(),
                CommonOptions.NameOption(),
                DiagnosticPortOption(),
                ShowChildIOOption()
            };
```

https://github.com/dotnet/diagnostics/blob/f2ee9569cfa90fa5c15fc27278785a55c555cb29/src/Tools/dotnet-trace/CommandLine/Commands/CollectCommand.cs#L380

```
        public static void Add(this Command command, ICommandHandler handler)
        {
            command.Handler = handler;
        }
```

https://github.com/dotnet/diagnostics/blob/f2ee9569cfa90fa5c15fc27278785a55c555cb29/src/Tools/Common/CommandExtensions.cs#L15

**Comment by huoyaoyuan** [comment]
The certain library usage is confusing, yes. But that feature isn't designed with that usage in mind. There's nothing prevent tricky usages in libraries.

**Reply by athinboy** [reply]
Frankly, I agree with you. It is just that  something is wrong .

**Comment by HaloFour** [comment]
I'm sorry that you do not appreciate the evolution of the language and find that it makes it difficult to comprehend.  Unfortunately C#, as almost all programming languages, is only going to continue to evolve and to add new features.  One tricky use of a feature does not make the feature bad.  Without extension methods LINQ would not have happened, and LINQ is probably of the most productivity-boosting features in the C# language.

**Reply by athinboy** [reply]
You are right . but ,tricky usage will be more and more . 
Back to this feature:
Collection initializers
Collection initializers let you specify one or more element initializers when you initialize a collection type that implements IEnumerable and has Add with the appropriate signature as an instance method or an extension method. 

Obviously,Command is a collection of Symbol class ,not ICommandHandler class, and ICommandHandler  is not sub class of 
Symbol class . so ,is this a precise Collection initializers?

```
    public class Command : Symbol, ICommand, ISymbol, ISuggestionSource, IEnumerable<Symbol>, IEnumerable
    {}
```


```
        public static void Add(this Command command, ICommandHandler handler)
        {
            command.Handler = handler;
        }
```

**Comment by jnm2** [comment]
> Last several years, csharplang had added more and more feature .

For what it's worth, the specific example you provide is of two features (object and collection initializers, extension methods) that were added 14 years ago in C# 3. (https://docs.microsoft.com/en-us/dotnet/csharp/whats-new/csharp-version-history#c-version-30)

**Reply by jnm2** [reply]
@athinboy Also, awesome avatar. You're the second person I've seen with that in dotnet/csharplang. :D

**Reply by athinboy** [reply]
I know collection initializers with {} ,just not agree with  extension method "Add" !

Command is a collection of Symbol class with a instance method "public void Add(Symbol symbol);" .


Collection initializers let you specify one or more element initializers when you initialize a collection type that implements IEnumerable and has Add with the appropriate signature as an instance method or an extension method.
